File: bblaser-service/src/main/java/nl/sijpesteijn/bblaserservice/converters/IldaConverter.java
package nl.sijpesteijn.bblaserservice.converters;

import nl.sijpesteijn.bblaserservice.domain.ilda.*;
import org.springframework.cglib.core.TinyBitSet;
import org.springframework.data.mongodb.core.aggregation.ArrayOperators;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class IldaConverter {
    private int coordinateHeaderSize = 32;
    private int coordinateDataSize = 8;
    private int colorHeaderSize = 32;
    private int colorDataSize = 6;
    private IldaColorTable colorDataTable = new IldaColorTable();
    private int maxHeight = 0;
    private int minHeight = 0;
    private int maxWidth = 0;
    private int minWidth = 0;
    private int maxDepth = 0;
    private int minDepth = 0;

    public Ilda fromBytes(final byte[] data) {
        List<ColorData> colorDataList = new ArrayList<>();
        ColorHeader colorHeader = null;
        int offset = 0;
        List<IldaFrame> frames = new ArrayList<>();
        IldaFrame frame = extractFrame(data, offset);
        frames.add(frame);

        while (frame.getTotalPoints() > 0) {
            offset += coordinateHeaderSize + (coordinateDataSize*frame.getTotalPoints());
            frame = extractFrame(data, offset);
            frames.add(frame);
        }

        Integer startColorHeader = offset + coordinateHeaderSize + (frame.getTotalPoints() * coordinateDataSize);

        if (startColorHeader < data.length) {
            byte[] dataBlock = Arrays.copyOfRange(data, startColorHeader, startColorHeader + colorHeaderSize);
            colorHeader = extractColorHeader(dataBlock);

            Integer startColorData = startColorHeader + colorHeaderSize;
            for (int i = 0; i < colorHeader.getTotalColors(); i++) {
                byte[] subDataBlock =  Arrays.copyOfRange(dataBlock, startColorData + i * colorDataSize,
                        startColorData + (i + 1) * colorDataSize);
                colorDataList.add(extractColorData(subDataBlock));
            }
        }

        return new Ilda(minWidth, maxWidth, minHeight, maxHeight, minDepth, maxDepth, colorHeader, frames, colorDataList);
    }

    private ColorData extractColorData(byte[] data) {
        Integer red1 = extractSmallInteger(data[0]);
        Integer red2 = extractSmallInteger(data[1]);
        Integer green1 = extractSmallInteger(data[2]);
        Integer green2 = extractSmallInteger(data[3]);
        Integer blue1 = extractSmallInteger(data[4]);
        Integer blue2 = extractSmallInteger(data[5]);
        return new ColorData(red1, red2, green1, green2, blue1, blue2, 12);
    }

    private IldaFrame extractFrame(byte[] data, int offset) {
        String protocol = "" + data[0 + offset] + data[1 + offset] + data[2 + offset] + data[3 + offset];
        Boolean threeD = data[7 + offset] == '0';
        String frameName = extractString(data, 8 + offset, 15 + offset);
        String companyName = extractString(data, 16 + offset, 23 + offset);

        Integer totalPoints = extractInteger(data[24 + offset], data[25 + offset]);
        Integer frameNumber = extractInteger(data[26 + offset], data[27 + offset]);
        Integer totalFrames = extractInteger(data[28 + offset], data[29 + offset]);
        Integer scannerHead = extractSmallInteger(data[30 + offset]);

        List<CoordinateData> coordinateDataList = new ArrayList<>();

        for (int i = 0; i < totalPoints; i++) {
            coordinateDataList.add(createCoordinateData(Arrays.copyOfRange(data, offset + coordinateHeaderSize + i * coordinateDataSize,
                    offset + coordinateHeaderSize + (i + 1) * coordinateDataSize)));
        }

        return new IldaFrame(protocol, threeD, frameName, companyName, totalPoints, frameNumber, totalFrames, scannerHead, coordinateDataList);
    }

    private CoordinateData createCoordinateData(byte[] dataBlock) {
        int x = extractCoordinate(dataBlock[0], dataBlock[1]);
        int y = extractCoordinate(dataBlock[2], dataBlock[3]);
        int z = extractCoordinate(dataBlock[4], dataBlock[5]);
        boolean blanked = isBitSet(dataBlock[6], 6);

        if (!blanked) {
            setMaxMinValue(x, y, z);
        }
        boolean endImageData = isBitSet(dataBlock[6], 7);
        Integer colorCode = extractSmallInteger(dataBlock[7]);
        ColorData color = colorDataTable.getColor(colorCode);

        return new CoordinateData(x, y, z, color, blanked, endImageData);
    }

    private boolean isBitSet(byte b, int bit) {
        return (b & (1 << bit)) != 0;
    }

    private int extractCoordinate(byte b1, byte b2) {
        return (b1 & 0xff) << 8 | (b2 & 0xff);
    }

    private ColorHeader extractColorHeader(byte[] data) {
        String protocol = "" + data[0] + data[1] + data[2] + data[3];
        Integer formatCode = extractSmallInteger(data[7]);
        String paletteName = extractString(data, 8, 15);
        String companyName = extractString(data, 16, 23);
        Integer totalColors = extractInteger(data[24], data[25]);
        Integer paletteNumber = extractInteger(data[26], data[27]);
        Integer scannerHead = extractSmallInteger(data[30]);
        return new ColorHeader(scannerHead, paletteNumber, totalColors, companyName, paletteName, formatCode, protocol);
    }

    private String extractString(byte[] data, int start, int end) {
        String str = "";
        for (int i = start; i < end; i++) {
            str += data[i];
        }
        return str;
    }

    private Integer extractInteger(byte b1, byte b2) {
        int high = b1 & 0xff;
        int low = b2 & 0xff;
        return (high << 8) | low;
    }

    private Integer extractSmallInteger(byte b) {
        return b & 0xff;
    }

    private void setMaxMinValue(int x, int y, int z) {
        if (x > maxWidth) {
            maxWidth = x;
        }
        if (x < minWidth) {
            minWidth = x;
        }
        if (y > maxHeight) {
            maxHeight = y;
        }
        if (y < minHeight) {
            minHeight = y;
        }
        if (z > maxDepth) {
            maxDepth = z;
        }
        if (z < minDepth) {
            minDepth = z;
        }
    }
}

File: bblaser-service/src/main/kotlin/nl/sijpesteijn/bblaser/ilda/IldaConverter.kt
package nl.sijpesteijn.bblaser.ilda

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class IldaConverter {
    private val coordinateHeaderSize = 32
    private val coordinateDataSize = 8
    private val colorHeaderSize = 32
    private val colorDataSize = 6
    private val colorDataTable = IldaColorTable()
    private var maxHeight = 0
    private var minHeight = 0
    private var maxWidth = 0
    private var minWidth = 0
    private var maxDepth = 0
    private var minDepth = 0


    fun read(data: ByteArray): Ilda {
        var colorHeader: ColorHeader? = null
        val colorDataList: MutableList<ColorData> = mutableListOf()
        var offset = 0
        val coordinateHeaderList = mutableListOf<IldaFrame>()
        var coordinateHeader = extractFrame(data, offset)
        coordinateHeaderList.add(coordinateHeader)

        while (coordinateHeader.totalPoints > 0) {
            offset += coordinateHeaderSize + (coordinateDataSize*coordinateHeader.totalPoints)
            coordinateHeader = extractFrame(data, offset)
            coordinateHeaderList.add(coordinateHeader)
        }

        val startColorHeader = offset + coordinateHeaderSize + (coordinateHeader.totalPoints * coordinateDataSize)

        if (startColorHeader < data.size) {
            val dataBlock = data.copyOfRange(startColorHeader, startColorHeader + colorHeaderSize)
            colorHeader = extractColorHeader(dataBlock)

            val startColoData = startColorHeader + colorHeaderSize
            for ( i in 0..colorHeader.totalColors) {
                val dataSubBlock = dataBlock.copyOfRange(startColoData + i * colorDataSize,
                                                         startColoData + (i + 1) * colorDataSize)
                colorDataList.add(extractColorData(dataSubBlock))
            }
        }

        return Ilda(
                minWidth,
                maxWidth,
                minHeight,
                maxHeight,
                minDepth,
                maxDepth,
                colorHeader,
                coordinateHeaderList,
                colorDataList
        )
    }

    private fun extractFrame(data: ByteArray, offset: Int): IldaFrame {
        val protocol = "" + data[0 + offset].toChar() + data[1 + offset].toChar() + data[2 + offset].toChar() + data[3 + offset].toChar()
        val threeD = data[7 + offset].toInt() == 0
        val frameName = extractString(data, 8 + offset, 15 + offset)
        val companyName = extractString(data, 16 + offset, 23 + offset)

        val totalPoints = extractInteger(data[24 + offset], data[25 + offset])
        val frameNumber = extractInteger(data[26 + offset], data[27 + offset])
        val totalFrames = extractInteger(data[28 + offset], data[29 + offset])
        val scannerHead = extractSmallInteger(data[30 + offset])

        val coordinateDataList = mutableListOf<CoordinateData>()
        for(i in 0..totalPoints) {
            val dataBlock = data.copyOfRange(offset + coordinateHeaderSize + i * coordinateDataSize,
                    offset + coordinateHeaderSize + (i + 1) * coordinateDataSize)
            coordinateDataList.add(createCoordinateData(dataBlock))
        }

        return IldaFrame(protocol, threeD, frameName, companyName, totalPoints, frameNumber, totalFrames, scannerHead, coordinateDataList)
    }

    private fun extractColorHeader(data: ByteArray): ColorHeader {
        val protocol = "" + data[0] + data[1] + data[2] + data[3]
        val formatCode = extractSmallInteger(data[7])
        val paletteName = extractString(data, 8, 15)
        val companyName = extractString(data, 16, 23)
        val totalColors = extractInteger(data[24], data[25])
        val paletteNumber = extractInteger(data[26], data[27])
        val scannerHead = extractSmallInteger(data[30])
        return ColorHeader(scannerHead, paletteNumber, totalColors, companyName, paletteName, formatCode, protocol)
    }

    private fun extractColorData(data: ByteArray): ColorData {
        val red1 = extractSmallInteger(data[0])
        val red2 = extractSmallInteger(data[1])
        val green1 = extractSmallInteger(data[2])
        val green2 = extractSmallInteger(data[3])
        val blue1 = extractSmallInteger(data[4])
        val blue2 = extractSmallInteger(data[5])
        return ColorData(red1, red2, green1, green2, blue1, blue2, 12)
    }

    private fun createCoordinateData(data: ByteArray): CoordinateData {
        val x = extractCoordinate(data[0], data[1])
        val y = extractCoordinate(data[2], data[3])
        val z = extractCoordinate(data[4], data[5])
        val blanked = isBitSet(data[6], 6)

        if (!blanked) {
            setMaxMinValue(x, y, z)
        }

        val endImageData = isBitSet(data[6], 7)
        val colorCode = extractSmallInteger(data[7])
        val color = colorDataTable.getColor(colorCode)
        return CoordinateData(x, y, x, color, blanked, endImageData)
    }

    private fun extractString(fileArray: ByteArray, start: Int, stop: Int): String {
        var name = ""
        for (i in start..stop) {
            name += fileArray[i].toChar()
        }
        return name
    }

    private fun extractInteger(b: Byte, b1: Byte): Int {
        val high = b.toInt() and 0xff
        val low = b1.toInt() and 0xff
        return high shl 8 or low
    }

    private fun extractSmallInteger(b: Byte): Int {
        return b.toInt() and 0xff
    }

    private fun extractCoordinate(b: Byte, b1: Byte): Int {
        return (b.toInt() and 0xFF shl 8 or (b1.toInt() and 0xFF)).toShort().toInt()
    }

    private fun setMaxMinValue(x: Int, y: Int, z: Int) {
        if (x > maxWidth) {
            maxWidth = x
        }
        if (x < minWidth) {
            minWidth = x
        }
        if (y > maxHeight) {
            maxHeight = y
        }
        if (y < minHeight) {
            minHeight = y
        }
        if (z > maxDepth) {
            maxDepth = z
        }
        if (z < minDepth) {
            minDepth = z
        }

    }

    private fun isBitSet(b: Byte, bit:Int): Boolean {
        return b.toInt() and (1 shl bit) != 0
    }
}

data class ColorHeader(val scannerHead: Int,
                       val paletteNumber: Int,
                       val totalColors: Int,
                       val companyName: String,
                       val paletteName: String,
                       val formatCode: Int,
                       val protocol: String)
data class ColorData(val red1: Int,
                     val red2: Int,
                     val green1: Int,
                     val green2: Int,
                     val blue1: Int,
                     val blue2: Int,
                     val code: Int)
data class CoordinateData(val x: Int,
                          val y: Int,
                          val z: Int,
                          val colorData: ColorData?,
                          val blanked: Boolean,
                          val endImageDate: Boolean)
data class IldaFrame(val protocol: String,
                     val threeD: Boolean,
                     val frameName: String,
                     val companyName: String,
                     val totalPoints: Int,
                     val frameNumber: Int,
                     val totalFrames: Int,
                     val scannerHead: Int,
                     var coordinateDataList: List<CoordinateData>?)
data class Ilda(val minWidth: Int,
                val maxWidth: Int,
                val minHeight: Int,
                val maxHeight: Int,
                val minDepth: Int,
                val maxDepth: Int,
                val colorHeader: ColorHeader?,
                val IldaFrames: List<IldaFrame>,
                val colorDataList: MutableList<ColorData>)

class IldaColorTable {
    private val colors = HashMap<Int, ColorData>()

    init {
        val inputStream = javaClass.getResourceAsStream("/ildaColorTable.csv")
        val reader = BufferedReader(InputStreamReader(inputStream))
        reader.lines().forEach {
            line ->
            val splits = line.split(" ")
            val code = splits[0].toInt()
            val red = splits[1].toInt()
            val green = splits[2].toInt()
            val blue = splits[3].toInt()
            val colorData = ColorData(red, red, green, green, blue, blue, code)
            colors[code] = colorData
        }
    }

    fun getColor(colorNr: Int): ColorData? {
        return colors[colorNr]
    }
}